File: nim/day25.nim
import math

const
  steps = 12667664
         # (write, move, state (A=0 .. F=5))
  states = [[(1,  1, 1), (0, -1, 2)],
            [(1, -1, 0), (1,  1, 3)],
            [(0, -1, 1), (0, -1, 4)],
            [(1,  1, 0), (0,  1, 1)],
            [(1, -1, 5), (1, -1, 2)],
            [(1,  1, 3), (1,  1, 0)]
           ]
var
  tape: array[-6000 .. 60, int] # tweaked after tracking min & max position
  position: int
  state = 0

for _ in 1 .. steps:
  let (write, move, newState) = states[state][tape[position]]
  tape[position] = write
  position += move
  state = newState

echo tape.sum

File: python/day25.py
from collections import defaultdict

steps = 12667664
states = {
    'A': ((1,  1, 'B'), (0, -1, 'C')),
    'B': ((1, -1, 'A'), (1,  1, 'D')),
    'C': ((0, -1, 'B'), (0, -1, 'E')),
    'D': ((1,  1, 'A'), (0,  1, 'B')),
    'E': ((1, -1, 'F'), (1, -1, 'C')),
    'F': ((1,  1, 'D'), (1,  1, 'A')),
}
tape = defaultdict(int)
position = 0
state = 'A'

for _ in range(steps):
    value = tape[position]
    write, move, state = states[state][value]
    tape[position] = write
    position += move

print(sum(tape.values()))
